tools/clang/test/HLSL/globallycoherent-template-errors.hlsl

```diff
@@ -2,18 +2,86 @@
 
 template <typename T> void doSomething(uint pos) {
   globallycoherent RWTexture2D<T> output;
-  globallycoherent Buffer<T> nonUAV; // expected-error {{'globallycoherent' is not a valid modifier for a non-UAV type}}
-  globallycoherent T ThisShouldBreak = 2.0; // expected-error {{'globallycoherent' is not a valid modifier for a non-UAV type}}
+  // expected-error@+2 {{'globallycoherent' is not a valid modifier for a declaration of type 'Buffer<float>'}}
+  // expected-note@+1 {{'globallycoherent' can only be applied to UAV or RWDispatchNodeInputRecord objects}}
+  globallycoherent Buffer<T> nonUAV;
+  // expected-error@+2 {{'globallycoherent' is not a valid modifier for a declaration of type 'float'}}
+  // expected-note@+1 {{'globallycoherent' can only be applied to UAV or RWDispatchNodeInputRecord objects}}
+  globallycoherent T ThisShouldBreak = 2.0;
   output[uint2(pos, pos)] = 0;
 }
 
 void doSomething2(uint pos) {
   globallycoherent RWTexture2D<float> output;
-  globallycoherent float ThisShouldBreak = 2.0; // expected-error {{'globallycoherent' is not a valid modifier for a non-UAV type}}
+  // expected-error@+2 {{'globallycoherent' is not a valid modifier for a declaration of type 'float'}}
+  // expected-note@+1 {{'globallycoherent' can only be applied to UAV or RWDispatchNodeInputRecord objects}}
+  globallycoherent float ThisShouldBreak = 2.0;
 }
 
 [numthreads(8, 8, 1)] void main(uint threadId
                                 : SV_DispatchThreadID) {
-  doSomething<float>(threadId); // expected-note {{in instantiation of function template specialization 'doSomething<float>' requested here}}
+  // expected-note@+1 {{in instantiation of function template specialization 'doSomething<float>' requested here}}
+  doSomething<float>(threadId);
   doSomething2(threadId);
 }
+
+template <typename T>
+void doSomething3(T data) {
+  T local1 = data;
+
+  // expected-warning@+1 {{implicit conversion from 'RWDispatchNodeInputRecord<Record>' to 'globallycoherent RWDispatchNodeInputRecord<Record>' adds globallycoherent annotation}}
+  globallycoherent T local2 = data;
+}
+
+template <typename T>
+void doSomething4(globallycoherent T data) {
+  // expected-warning@+1 {{implicit conversion from 'globallycoherent RWDispatchNodeInputRecord<Record>' to 'RWDispatchNodeInputRecord<Record>' loses globallycoherent annotation}}
+  doSomething3(data);
+
+  // expected-warning@+1 {{implicit conversion from 'globallycoherent RWDispatchNodeInputRecord<Record>' to 'RWDispatchNodeInputRecord<Record>' loses globallycoherent annotation}}
+  T local1 = data;
+
+  globallycoherent T local2 = data;
+
+  doSomething3(local1);
+}
+
+// TBD: Support node types in templates.
+//  Currently, these produce:
+//  'T' cannot be used as a type parameter where a struct/class is required
+//  Is this #5729?
```

Comment:
I filed issue #5822 for this.

Reply:
Since template fixes are going in, I might wait until those merge, then update this test.

Reply:
Looks like I need to wait a bit longer, so I'll merge this and circle back once these are supported.